providers/class.two-factor-backup-codes.php

```diff
@@ -0,0 +1,274 @@
+							$( '.two-factor-backup-codes-wrapper' ).show();
+							$( '.two-factor-backup-codes-unused-codes' ).html( '' );
+							$.each( data, function( key, val ) {
+								$( '.two-factor-backup-codes-unused-codes' ).append( '<li>' + val + '</li>' );
+							} );
+							// Update counter.
+							$( '.two-factor-backup-codes-count' ).html( data.length );
+						},
+						error: function( errorThrown ) {
+							alert( 'error' ); // @todo: remove for production
+							console.log( errorThrown );
+						}
+					} );
+				} );
+			} );
+		</script>
+		<?php
+	}
+
+	/**
+	 * Generates backup codes & updates the user meta.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int $user_id The logged-in user ID.
+	 */
+	public function generate_codes( $user_id ) {
+		$codes = array();
+		$codes_hashed = array();
+
+		for ( $i = 0; $i < self::NUMBER_OF_CODES; $i++ ) {
+			$code = $this->get_code();
+			$codes_hashed[] = wp_hash_password( $code );
+			$codes[] = $code;
+			unset( $code );
+		}
+
+		update_user_meta( $user_id, self::BACKUP_CODES_META_KEY, $codes_hashed );
+
+		// Unhashed.
+		return $codes;
+	}
+
+	/**
+	 * Generates a JSON object of backup codes.
+	 *
+	 * @since 0.1-dev
+	 */
+	public function ajax_generate_json() {
+		check_ajax_referer( 'two-factor-backup-codes-generate-json', 'nonce' );
+
+		$user_id = get_current_user_id();
+		$codes = $this->generate_codes( $user_id );
+		wp_send_json( $codes );
+	}
+
+	/**
+	 * Prints the form that prompts the user to authenticate.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function authentication_page( $user ) {
+		require_once( ABSPATH .  '/wp-admin/includes/template.php' );
+		?>
+		<p><?php esc_html_e( 'Enter a backup verification code.' ); ?></p><br/>
+		<p>
+			<label for="authcode"><?php esc_html_e( 'Verification Code:' ); ?></label>
+			<input type="tel" name="two-factor-backup-code" id="authcode" class="input" value="" size="20" pattern="[0-9]*" />
+		</p>
+		<?php
+		submit_button( __( 'Submit' ) );
+	}
+
+	/**
+	 * Validates the users input token.
+	 *
+	 * In this class we just return true.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 * @return boolean
+	 */
+	public function validate_authentication( $user ) {
+		return $this->validate_code( $user->ID, $_REQUEST['two-factor-backup-code'] );
+	}
+
+	/**
+	 * Validates a backup code.
+	 *
+	 * Backup Codes are single use and are deleted upon a successful validation.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int $user_id The logged-in user ID.
+	 * @param int $code    The backup code.
+	 * @return boolean
+	 */
+	public function validate_code( $user_id, $code ) {
+		$backup_codes = get_user_meta( $user_id, self::BACKUP_CODES_META_KEY, true );
+
+		foreach ( $backup_codes as $code_index => $backup_code ) {
+			if ( wp_check_password( $code, $backup_code, $user_id ) ) {
+				$this->delete_code( $user_id, $code_index );
```

Comment:
I'd sooner see this delete by the `$backup_code` hashed version than array index.  Array index feels like a relatively arbitrary index to use.